File: apigen/test/BuiltinsTest.java
package test;

import test.builtins.*;

public class BuiltinsTest
{

  private BuiltinsFactory factory;
  static private String directory;

  public BuiltinsTest(BuiltinsFactory factory) {
    this.factory = factory;
  }

  public void run() throws java.io.IOException {
  
      T t = factory.makeT_Trm(factory.parse("one"));
      testAssert(t.toString().equals("term(one)"), "make term test");
      
      D d = factory.makeD_Double(new Double(1.0));
      testAssert(d.toString().equals("double(1.0)"),"make double test");
      
      I i = factory.makeI_Integer(new Integer(1));
      testAssert(i.toString().equals("int(1)"),"make int test");
      
      L l = factory.makeL_Lst((aterm.ATermList) factory.parse("[one]"));
      testAssert(l.toString().equals("list([one])"), "make list test");
      
      S s = factory.makeS_String("one");
      testAssert(s.toString().equals("str(\"one\")"), "make str test");
  }

  public final static void main(String[] args) throws jjtraveler.VisitFailure,
  java.io.IOException
  {
    BuiltinsTest test = new BuiltinsTest (new BuiltinsFactory()); 
    directory = args[0];

    test.run();
    return;
  }

  void testAssert(boolean b, String name) {
    if (!b) {
      throw new RuntimeException("Test " + name + " failed!");
    }
  }
}

File: apigen/test/builtinstest.c
#include <stdio.h>
#include <assert.h>

#include "builtins.h"

int main(int argc, char *argv[])
{
  ATerm bottomOfStack;

  ATinit(argc, argv, &bottomOfStack);
  initBuiltinsApi();

  ATerm data[20];

  /* Note that there is still a bug in ApiGen,
   * If all of these constructors were in the same sort,
   * wrong code would have been generated. This needs
   * to be fixed!
   */

  data[0] = (ATerm) makeIInteger(1);
  assert(data[0] && ATisEqual(data[0], ATparse("int(1)")));

  data[1] = (ATerm) makeDDouble(1.0);
  assert(data[1] && ATisEqual(data[1], ATparse("double(1.0)")));

  data[2] = (ATerm) makeSString("one");
  assert(data[2] && ATisEqual(data[2], ATparse("str(\"one\")")));

  data[3] = (ATerm) makeTTrm(ATparse("one"));
  assert(data[3] && ATisEqual(data[3], ATparse("term(one)")));

  data[4] = (ATerm) makeLLst((ATermList) ATparse("[one]"));
  assert(data[4] && ATisEqual(data[4], ATparse("list([one])")));


  return 0;
}
